From 6cd1322c2f50be5e5bb6206e52e45d6835827f58 Mon Sep 17 00:00:00 2001
Message-ID: <6cd1322c2f50be5e5bb6206e52e45d6835827f58.1779975966.git.bertrand.marquis@arm.com>
In-Reply-To: <55d56a174a0f04520d4a1558630bb0b739074b13.1779975966.git.bertrand.marquis@arm.com>
References: <55d56a174a0f04520d4a1558630bb0b739074b13.1779975966.git.bertrand.marquis@arm.com>
From: Bertrand Marquis <bertrand.marquis@arm.com>
Date: Fri, 22 May 2026 12:04:28 +0200
Subject: [PATCH 18/30] virtio-msg: Report missing devices as transport status

Make virtio-msg device lookup return explicit transport status codes
instead of logging and continuing with fabricated zeroed responses.

Return VIRTIO_MSG_ERROR_NO_DEVICE when a valid device slot has no
attached virtio device, while invalid slots remain bad peer messages.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>
---
 hw/virtio/virtio-msg.c              | 185 ++++++++++++++++++----------
 include/hw/virtio/virtio-msg-prot.h |   1 +
 2 files changed, 123 insertions(+), 63 deletions(-)

diff --git a/hw/virtio/virtio-msg.c b/hw/virtio/virtio-msg.c
index 9ae0d0e042a8..12c918391cab 100644
--- a/hw/virtio/virtio-msg.c
+++ b/hw/virtio/virtio-msg.c
@@ -168,23 +168,39 @@ static bool virtio_msg_bus_bad(VirtIOMSGProxy *s, VirtIOMSG *msg)
 
 static VirtIODevice *virtio_msg_vdev(VirtIOMSGProxy *s, uint16_t dev_num)
 {
-    VirtIODevice *vdev;
+    uint16_t slot;
+
+    if (!virtio_msg_dev_num_to_slot(s, dev_num, &slot)) {
+        return NULL;
+    }
 
-    vdev = virtio_bus_get_device(&s->devs[dev_num].bus);
-    return vdev;
+    return virtio_bus_get_device(&s->devs[slot].bus);
 }
 
-static VirtIODevice *virtio_msg_lookup_vdev(VirtIOMSGProxy *s, uint16_t dev_num,
-                                            const char *what)
+static int virtio_msg_lookup_vdev(VirtIOMSGProxy *s, uint16_t dev_num,
+                                  uint16_t *slot, VirtIODevice **vdev)
 {
-    VirtIODevice *vdev = virtio_msg_vdev(s, dev_num);
+    uint16_t target_slot;
+    VirtIODevice *target_vdev;
 
-    if (!vdev) {
-        error_report("%s: No virtio device on bus %s!",
-                     what, BUS(&s->devs[dev_num].bus)->name);
+    if (!virtio_msg_dev_num_to_slot(s, dev_num, &target_slot)) {
+        return VIRTIO_MSG_ERROR_BAD_MESSAGE;
     }
 
-    return vdev;
+    target_vdev = virtio_bus_get_device(&s->devs[target_slot].bus);
+
+    if (!target_vdev) {
+        return VIRTIO_MSG_ERROR_NO_DEVICE;
+    }
+
+    if (slot) {
+        *slot = target_slot;
+    }
+    if (vdev) {
+        *vdev = target_vdev;
+    }
+
+    return VIRTIO_MSG_NO_ERROR;
 }
 
 static int virtio_msg_bus_get_devices(VirtIOMSGProxy *s,
@@ -226,22 +242,23 @@ static int virtio_msg_bus_ping(VirtIOMSGProxy *s, VirtIOMSG *msg)
 static int virtio_msg_device_info(VirtIOMSGProxy *s,
                                   VirtIOMSG *msg)
 {
-    VirtIODevice *vdev = virtio_msg_vdev(s, msg->dev_num);
+    VirtIODevice *vdev;
     uint32_t config_len = 0;
     uint32_t device_id = 0;
     uint32_t max_virtqueues = 0;
     uint8_t nil_uuid[16] = { 0 };
     VirtIOMSG msg_resp;
+    int ret;
 
-    if (vdev) {
-        device_id = vdev->device_id;
-        config_len = vdev->config_len;
-        max_virtqueues = virtio_get_num_queues(vdev);
-    } else {
-        error_report("%s: No virtio device on bus %s!",
-                     __func__, BUS(&s->devs[msg->dev_num].bus)->name);
+    ret = virtio_msg_lookup_vdev(s, msg->dev_num, NULL, &vdev);
+    if (ret != VIRTIO_MSG_NO_ERROR) {
+        return ret;
     }
 
+    device_id = vdev->device_id;
+    config_len = vdev->config_len;
+    max_virtqueues = virtio_get_num_queues(vdev);
+
     virtio_msg_pack_get_device_info_resp(&msg_resp, msg->dev_num, msg->token,
                                          device_id,
                                          VIRTIO_MSG_VENDOR_ID,
@@ -256,20 +273,23 @@ static int virtio_msg_device_info(VirtIOMSGProxy *s,
 static int virtio_msg_get_features(VirtIOMSGProxy *s,
                                    VirtIOMSG *msg)
 {
-    VirtIODevice *vdev = virtio_msg_lookup_vdev(s, msg->dev_num, __func__);
+    VirtIODevice *vdev;
     VirtIOMSG msg_resp;
     uint32_t index = msg->get_features.block_index;
     uint32_t f[VIRTIO_MSG_MAX_FEATURE_NUM] = { 0 };
     uint32_t num = MIN(msg->get_features.num_blocks,
                        VIRTIO_MSG_MAX_FEATURE_NUM);
     uint64_t features = 0;
+    int ret;
 
-    if (vdev) {
-        VirtioDeviceClass *vdc = VIRTIO_DEVICE_GET_CLASS(vdev);
-
-        features = vdev->host_features & ~vdc->legacy_features;
+    ret = virtio_msg_lookup_vdev(s, msg->dev_num, NULL, &vdev);
+    if (ret != VIRTIO_MSG_NO_ERROR) {
+        return ret;
     }
 
+    features = vdev->host_features &
+               ~VIRTIO_DEVICE_GET_CLASS(vdev)->legacy_features;
+
     /* We only have 64 feature bits. If driver asks for more, return zeros  */
     if (index < 2) {
         features >>= index * 32;
@@ -289,8 +309,15 @@ static int virtio_msg_set_features(VirtIOMSGProxy *s,
     VirtIOMSG msg_resp;
     unsigned int i;
     uint64_t f;
+    uint16_t slot;
+    int ret;
+
+    ret = virtio_msg_lookup_vdev(s, msg->dev_num, &slot, NULL);
+    if (ret != VIRTIO_MSG_NO_ERROR) {
+        return ret;
+    }
 
-    f = s->devs[msg->dev_num].guest_features;
+    f = s->devs[slot].guest_features;
 
     for (i = 0; i < msg->set_features.num_blocks; i++) {
         unsigned int feature_index = i + msg->set_features.block_index;
@@ -304,7 +331,7 @@ static int virtio_msg_set_features(VirtIOMSGProxy *s,
                       msg->set_features.features[i]);
     }
 
-    s->devs[msg->dev_num].guest_features = f;
+    s->devs[slot].guest_features = f;
 
     virtio_msg_pack_set_features_resp(&msg_resp, msg->dev_num, msg->token);
     return virtio_msg_send_must_deliver_checked(s, &msg_resp);
@@ -321,31 +348,34 @@ static void virtio_msg_soft_reset(VirtIOMSGProxy *s, uint16_t dev_num)
 static int virtio_msg_set_device_status(VirtIOMSGProxy *s,
                                         VirtIOMSG *msg)
 {
-    VirtIODevice *vdev = virtio_msg_vdev(s, msg->dev_num);
+    uint16_t slot;
+    VirtIODevice *vdev;
     uint32_t status = msg->set_device_status.status;
     VirtIOMSG msg_resp;
+    int ret;
 
-    if (!vdev) {
-        return VIRTIO_MSG_NO_ERROR;
+    ret = virtio_msg_lookup_vdev(s, msg->dev_num, &slot, &vdev);
+    if (ret != VIRTIO_MSG_NO_ERROR) {
+        return ret;
     }
 
     if (!(status & VIRTIO_CONFIG_S_DRIVER_OK)) {
-        virtio_bus_stop_ioeventfd(&s->devs[msg->dev_num].bus);
+        virtio_bus_stop_ioeventfd(&s->devs[slot].bus);
     }
 
     if (status & VIRTIO_CONFIG_S_FEATURES_OK) {
-        virtio_set_features(vdev, s->devs[msg->dev_num].guest_features);
+        virtio_set_features(vdev, s->devs[slot].guest_features);
     }
 
     virtio_set_status(vdev, status);
     assert(vdev->status == status);
 
     if (status & VIRTIO_CONFIG_S_DRIVER_OK) {
-        virtio_bus_start_ioeventfd(&s->devs[msg->dev_num].bus);
+        virtio_bus_start_ioeventfd(&s->devs[slot].bus);
     }
 
     if (status == 0) {
-        virtio_msg_soft_reset(s, msg->dev_num);
+        virtio_msg_soft_reset(s, slot);
     }
 
     virtio_msg_pack_set_device_status_resp(&msg_resp, msg->dev_num, msg->token,
@@ -356,25 +386,31 @@ static int virtio_msg_set_device_status(VirtIOMSGProxy *s,
 static int virtio_msg_get_device_status(VirtIOMSGProxy *s,
                                         VirtIOMSG *msg)
 {
-    VirtIODevice *vdev = virtio_msg_lookup_vdev(s, msg->dev_num, __func__);
+    VirtIODevice *vdev;
     VirtIOMSG msg_resp;
-    uint32_t status = vdev ? vdev->status : 0;
+    int ret;
+
+    ret = virtio_msg_lookup_vdev(s, msg->dev_num, NULL, &vdev);
+    if (ret != VIRTIO_MSG_NO_ERROR) {
+        return ret;
+    }
 
     virtio_msg_pack_get_device_status_resp(&msg_resp, msg->dev_num, msg->token,
-                                           status);
+                                           vdev->status);
     return virtio_msg_send_must_deliver_checked(s, &msg_resp);
 }
 
 static int virtio_msg_get_config(VirtIOMSGProxy *s,
                                  VirtIOMSG *msg)
 {
-    VirtIODevice *vdev = virtio_msg_lookup_vdev(s, msg->dev_num, __func__);
+    VirtIODevice *vdev;
     uint32_t size = msg->get_config.size;
     uint32_t offset = msg->get_config.offset;
     uint8_t data[VIRTIO_MSG_MAX_CONFIG_BYTES];
     VirtIOMSG msg_resp;
     uint32_t generation = 0;
     unsigned int i;
+    int ret;
 
     if (size > VIRTIO_MSG_MAX_CONFIG_BYTES) {
         return VIRTIO_MSG_NO_ERROR;
@@ -382,13 +418,16 @@ static int virtio_msg_get_config(VirtIOMSGProxy *s,
 
     memset(data, 0, size);
 
-    if (vdev) {
-        for (i = 0; i < size; i++) {
-            data[i] = virtio_config_modern_readb(vdev, offset + i);
-        }
-        generation = vdev->generation;
+    ret = virtio_msg_lookup_vdev(s, msg->dev_num, NULL, &vdev);
+    if (ret != VIRTIO_MSG_NO_ERROR) {
+        return ret;
     }
 
+    for (i = 0; i < size; i++) {
+        data[i] = virtio_config_modern_readb(vdev, offset + i);
+    }
+    generation = vdev->generation;
+
     virtio_msg_pack_get_config_resp(&msg_resp, msg->dev_num, msg->token,
                                     size, offset,
                                     generation, data);
@@ -398,20 +437,24 @@ static int virtio_msg_get_config(VirtIOMSGProxy *s,
 static int virtio_msg_set_config(VirtIOMSGProxy *s,
                                  VirtIOMSG *msg)
 {
-    VirtIODevice *vdev = virtio_msg_lookup_vdev(s, msg->dev_num, __func__);
+    VirtIODevice *vdev;
     uint32_t offset = msg->set_config.offset;
     uint32_t size = msg->set_config.size;
     uint8_t *data = msg->set_config.data;
     VirtIOMSG msg_resp;
     uint32_t generation = 0;
     unsigned int i;
+    int ret;
 
-    if (vdev) {
-        for (i = 0; i < size; i++) {
-            virtio_config_modern_writeb(vdev, offset + i, data[i]);
-        }
-        generation = vdev->generation;
+    ret = virtio_msg_lookup_vdev(s, msg->dev_num, NULL, &vdev);
+    if (ret != VIRTIO_MSG_NO_ERROR) {
+        return ret;
+    }
+
+    for (i = 0; i < size; i++) {
+        virtio_config_modern_writeb(vdev, offset + i, data[i]);
     }
+    generation = vdev->generation;
 
     virtio_msg_pack_set_config_resp(&msg_resp, msg->dev_num, msg->token,
                                     size, offset,
@@ -422,14 +465,20 @@ static int virtio_msg_set_config(VirtIOMSGProxy *s,
 static int virtio_msg_get_vqueue(VirtIOMSGProxy *s,
                                  VirtIOMSG *msg)
 {
-    VirtIODevice *vdev = virtio_msg_lookup_vdev(s, msg->dev_num, __func__);
+    VirtIODevice *vdev;
     uint32_t max_size = 0;
     uint32_t index = msg->get_vqueue.index;
     hwaddr desc = 0, avail = 0, used = 0;
     VirtIOMSG msg_resp;
     uint32_t size = 0;
+    int ret;
+
+    ret = virtio_msg_lookup_vdev(s, msg->dev_num, NULL, &vdev);
+    if (ret != VIRTIO_MSG_NO_ERROR) {
+        return ret;
+    }
 
-    if (index < VIRTIO_QUEUE_MAX && vdev) {
+    if (index < VIRTIO_QUEUE_MAX) {
         max_size = virtio_queue_get_max_num(vdev, index);
         size = virtio_queue_get_num(vdev, index);
         if (size) {
@@ -439,7 +488,7 @@ static int virtio_msg_get_vqueue(VirtIOMSGProxy *s,
                                         index, max_size, size, 0,
                                         desc, avail, used);
     } else {
-        /* OOB index or missing device, respond with all zeroes. */
+        /* OOB index, respond with all zeroes. */
         virtio_msg_pack_get_vqueue_resp(&msg_resp, msg->dev_num, msg->token,
                                         index, 0, 0, 0, 0, 0, 0);
     }
@@ -449,12 +498,18 @@ static int virtio_msg_get_vqueue(VirtIOMSGProxy *s,
 
 static int virtio_msg_set_vqueue(VirtIOMSGProxy *s, VirtIOMSG *msg)
 {
-    VirtIODevice *vdev = virtio_msg_lookup_vdev(s, msg->dev_num, __func__);
+    VirtIODevice *vdev;
     uint32_t index = msg->set_vqueue.index;
     VirtIOMSG msg_resp;
+    int ret;
+
+    ret = virtio_msg_lookup_vdev(s, msg->dev_num, NULL, &vdev);
+    if (ret != VIRTIO_MSG_NO_ERROR) {
+        return ret;
+    }
 
-    if (!vdev || index >= VIRTIO_QUEUE_MAX) {
-        /* Missing device or OOB index, ignore. */
+    if (index >= VIRTIO_QUEUE_MAX) {
+        /* OOB index, ignore. */
         return VIRTIO_MSG_NO_ERROR;
     }
 
@@ -472,11 +527,13 @@ static int virtio_msg_set_vqueue(VirtIOMSGProxy *s, VirtIOMSG *msg)
 
 static int virtio_msg_reset_vqueue(VirtIOMSGProxy *s, VirtIOMSG *msg)
 {
-    VirtIODevice *vdev = virtio_msg_lookup_vdev(s, msg->dev_num, __func__);
+    VirtIODevice *vdev;
     VirtIOMSG msg_resp;
+    int ret;
 
-    if (!vdev) {
-        return VIRTIO_MSG_NO_ERROR;
+    ret = virtio_msg_lookup_vdev(s, msg->dev_num, NULL, &vdev);
+    if (ret != VIRTIO_MSG_NO_ERROR) {
+        return ret;
     }
 
     virtio_queue_reset(vdev, msg->reset_vqueue.index);
@@ -488,12 +545,14 @@ static int virtio_msg_reset_vqueue(VirtIOMSGProxy *s, VirtIOMSG *msg)
 static int virtio_msg_event_avail(VirtIOMSGProxy *s,
                                   VirtIOMSG *msg)
 {
-    VirtIODevice *vdev = virtio_msg_lookup_vdev(s, msg->dev_num, __func__);
+    VirtIODevice *vdev;
     uint16_t vq_idx = msg->event_avail.index;
     VirtQueue *vq;
+    int ret;
 
-    if (!vdev) {
-        return VIRTIO_MSG_NO_ERROR;
+    ret = virtio_msg_lookup_vdev(s, msg->dev_num, NULL, &vdev);
+    if (ret != VIRTIO_MSG_NO_ERROR) {
+        return ret;
     }
 
     if (!(vdev->status & VIRTIO_CONFIG_S_DRIVER_OK)) {
@@ -604,7 +663,7 @@ static void virtio_msg_notify(DeviceState *opaque, uint16_t vector)
 {
     VirtIOMSGDev *mdev = VIRTIO_MSG_DEV(opaque);
     VirtIOMSGProxy *s = VIRTIO_MSG(mdev->proxy);
-    VirtIODevice *vdev = virtio_msg_lookup_vdev(s, mdev->dev_num, __func__);
+    VirtIODevice *vdev = virtio_msg_vdev(s, mdev->dev_num);
     VirtIOMSG msg;
 
     if (!vdev || !virtio_msg_bus_connected(&s->msg_bus)) {
@@ -821,7 +880,7 @@ static int virtio_msg_set_guest_notifier(DeviceState *d, int n, bool assign,
 {
     VirtIOMSGDev *mdev = VIRTIO_MSG_DEV(d);
     VirtIOMSGProxy *s = VIRTIO_MSG(mdev->proxy);
-    VirtIODevice *vdev = virtio_msg_lookup_vdev(s, mdev->dev_num, __func__);
+    VirtIODevice *vdev = virtio_msg_vdev(s, mdev->dev_num);
     VirtQueue *vq;
     EventNotifier *notifier;
     VirtioDeviceClass *vdc;
@@ -857,7 +916,7 @@ static int virtio_msg_set_config_guest_notifier(DeviceState *d, bool assign,
 {
     VirtIOMSGDev *mdev = VIRTIO_MSG_DEV(d);
     VirtIOMSGProxy *s = VIRTIO_MSG(mdev->proxy);
-    VirtIODevice *vdev = virtio_msg_lookup_vdev(s, mdev->dev_num, __func__);
+    VirtIODevice *vdev = virtio_msg_vdev(s, mdev->dev_num);
     EventNotifier *notifier;
     VirtioDeviceClass *vdc;
     int r = 0;
@@ -890,7 +949,7 @@ static int virtio_msg_set_guest_notifiers(DeviceState *d, int nvqs,
 {
     VirtIOMSGDev *mdev = VIRTIO_MSG_DEV(d);
     VirtIOMSGProxy *s = VIRTIO_MSG(mdev->proxy);
-    VirtIODevice *vdev = virtio_msg_lookup_vdev(s, mdev->dev_num, __func__);
+    VirtIODevice *vdev = virtio_msg_vdev(s, mdev->dev_num);
     /* Mirror virtio-mmio: use eventfd handlers and skip irqfd for now. */
     bool with_irqfd = false;
     int r, n;
diff --git a/include/hw/virtio/virtio-msg-prot.h b/include/hw/virtio/virtio-msg-prot.h
index ce397968e939..18bdb154174f 100644
--- a/include/hw/virtio/virtio-msg-prot.h
+++ b/include/hw/virtio/virtio-msg-prot.h
@@ -19,6 +19,7 @@ enum {
     VIRTIO_MSG_ERROR_TIMEOUT,
     VIRTIO_MSG_ERROR_UNSUPPORTED_MESSAGE_ID,
     VIRTIO_MSG_ERROR_BAD_MESSAGE,
+    VIRTIO_MSG_ERROR_NO_DEVICE,
     VIRTIO_MSG_ERROR_MEMORY, /* General memory error. */
 };
 
-- 
2.53.0

